File: src/extensions/resource.py
# -*- mode: python; encoding: utf-8 -*-
#
# Copyright 2012 Jens Lindström, Opera Software ASA
#
# Licensed under the Apache License, Version 2.0 (the "License"); you may not
# use this file except in compliance with the License.  You may obtain a copy of
# the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
# License for the specific language governing permissions and limitations under
# the License.

import os
import errno

import configuration

from extensions import getExtensionPath, getExtensionInstallPath

def get(req, db, user, path):
    extension_name, resource_path = path.split("/", 1)

    cursor = db.cursor()
    cursor.execute("""SELECT users.name, extensionversions.sha1
                        FROM extensions
                        JOIN extensioninstalls ON (extensioninstalls.extension=extensions.id)
             LEFT OUTER JOIN extensionversions ON (extensionversions.id=extensioninstalls.version)
             LEFT OUTER JOIN users ON (users.id=extensions.author)
                       WHERE extensions.name=%s
                         AND (extensioninstalls.uid=%s OR extensioninstalls.uid IS NULL)
                    ORDER BY extensioninstalls.uid ASC NULLS LAST
                       LIMIT 1""",
                   (extension_name, user.id))

    row = cursor.fetchone()

    if not row:
        return None, None

    author_name, version_sha1 = row

    if version_sha1 is None:
        extension_path = getExtensionPath(author_name, extension_name)
    else:
        extension_path = getExtensionInstallPath(version_sha1)

    resource_path = os.path.join(extension_path, "resources", resource_path)

    def guessContentType(name):
        try:
            name, ext = name.split(".", 1)
            return configuration.mimetypes.MIMETYPES[ext]
        except:
            return "application/octet-stream"

    try:
        with open(resource_path) as resource_file:
            resource = resource_file.read()
    except IOError as error:
        if error.errno in (errno.ENOENT, errno.EACCES):
            return None, None
        raise
    else:
        return guessContentType(resource_path), resource

File: ui/src/reducers/resource.ts
/*
 * Copyright 2019 the Critic contributors, Opera Software ASA
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License.  You may obtain a copy
 * of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import produce from "./immer"
import { Reducer } from "redux"

import { DATA_UPDATE, Action } from "../actions"

interface ResourceWithID<ID> {
  id: ID
}

function defaultGetID<ID, T extends ResourceWithID<ID>>(resource: T): ID {
  return resource.id
}

//type Reducer<State> = (state: State, action: Action) => State

/* Reducer for the primary mapping of resoucre ID to resource object. */
export const primaryMap = <T, ID extends number | string>(
  resourceName: string,
  getID: (resource: T) => ID | null = defaultGetID as (resource: T) => ID
) =>
  produce<Map<ID, T>>((draft: Map<ID, T>, action: Action) => {
    if (action.type === DATA_UPDATE) {
      const updates = action.updates.get(resourceName) as T[]
      if (updates)
        for (const item of updates) {
          const id = getID(item)
          if (id !== null) draft.set(id, item)
        }
      const deleted = action.deleted?.get(resourceName) as Set<ID> | undefined
      if (deleted) for (const id of deleted) draft.delete(id)
    }
  }, new Map())

const lookupDelete = <Key, ID>(draft: Map<Key, ID>, deleted?: Set<ID>) => {
  if (deleted) {
    const deleteKeys = [...draft.entries()]
      .filter(([key, id]) => deleted.has(id))
      .map(([key]) => key)
    for (const key of deleteKeys) draft.delete(key)
  }
}

/* Reducer for lookup mappings of secondary key to resource ID. */
export const lookupMap = <T, Key, ID extends number | string>(
  resourceName: string,
  getKey: (resource: T) => null | Key,
  getID: (resource: T) => ID = defaultGetID as (resource: T) => ID
) =>
  produce<Map<Key, ID>>((draft: Map<Key, ID>, action: Action) => {
    if (action.type === DATA_UPDATE) {
      const updates = action.updates.get(resourceName) as T[]
      if (updates)
        for (const item of updates) {
          const key = getKey(item)
          if (key !== null) draft.set(key, getID(item))
        }
      lookupDelete(
        draft,
        action.deleted?.get(resourceName) as Set<ID> | undefined
      )
    }
  }, new Map())

/* Reducer for lookup mappings of multiple secondary keys to resource ID. */
export const lookupManyMap = <T, Key, ID extends number | string>(
  resourceName: string,
  getKeys: (resource: T) => Key[],
  getID: (resource: T) => ID = defaultGetID as (resource: T) => ID
) =>
  produce<Map<Key, ID>>((draft: Map<Key, ID>, action: Action) => {
    if (action.type === DATA_UPDATE) {
      const updates = action.updates.get(resourceName) as T[]
      if (updates)
        for (const item of updates) {
          const id = getID(item)
          for (const key of getKeys(item)) draft.set(key, id)
        }
      lookupDelete(
        draft,
        action.deleted?.get(resourceName) as Set<ID> | undefined
      )
    }
  }, new Map())

/* Reducer for mappings of resource ID (or other key) to auxilliary data. */
export const auxilliaryMap = <T, ID, Auxilliary>(
  resourceName: string,
  mapper: (resource: T) => [ID, Auxilliary | null]
) =>
  produce<Map<ID, Auxilliary>>((draft: Map<ID, Auxilliary>, action: Action) => {
    if (action.type === DATA_UPDATE) {
      const updates = action.updates.get(resourceName)
      if (updates)
        for (const item of updates) {
          const [id, auxilliary] = mapper(item)
          if (auxilliary) draft.set(id, auxilliary)
        }
      const deleted = action.deleted?.get(resourceName) as Set<ID> | undefined
      if (deleted) for (const id of deleted) draft.delete(id)
    }
  }, new Map())
